File: 3Sum Closest - Leetcode 16/3Sum Closest - Leetcode 16.cpp
#include <vector>
#include <algorithm>
#include <cstdlib>

class Solution {
public:
    int threeSumClosest(std::vector<int>& nums, int target) {
        int n = nums.size();
        std::sort(nums.begin(), nums.end());
        int closestSum = INT_MAX;

        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int lo = i + 1, hi = n - 1;
            while (lo < hi) {
                int curSum = nums[i] + nums[lo] + nums[hi];

                if (std::abs(curSum - target) < std::abs(closestSum - target)) {
                    closestSum = curSum;
                }

                if (curSum == target) {
                    return curSum;
                } else if (curSum < target) {
                    lo++;
                } else {
                    hi--;
                }
            }
        }
        return closestSum;
    }
};

File: 3Sum Closest - Leetcode 16/3Sum Closest - Leetcode 16.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        closest_sum = float('inf')

        for i in range(n):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            lo, hi = i + 1, n - 1
            while lo < hi:
                cur_sum = nums[i] + nums[lo] + nums[hi]

                if abs(cur_sum - target) < abs(closest_sum - target):
                    closest_sum = cur_sum

                if cur_sum == target:
                    return cur_sum
                elif cur_sum < target:
                    lo += 1
                else:
                    hi -= 1

        return closest_sum
        # Time: O(n^2)
        # Space: O(n)
